### PR TITLE
NWM_UDS: implement disconnect_reason and EjectClient

### PR DESCRIPTION
**Pull request by B3n30**
This implements the EjectClient function. It is used by some games to disconnect a client from the session

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/5331)

### CODE REVIEW

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 1]
```diff
@@ -928,6 +931,57 @@ void NWM_UDS::BeginHostingNetworkDeprecated(Kernel::HLERequestContext& ctx) {
     rb.Push(result);
 }
 
+void NWM_UDS::EjectClient(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x05, 1, 0);
+    u16 network_node_id = rp.Pop<u16>();
```
Comment (lioncash) [inline src/core/hle/service/nwm/nwm_uds.cpp:936]:
```suggestion
    const auto network_node_id = rp.Pop<u16>();
```

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 2]
```diff
@@ -928,6 +931,57 @@ void NWM_UDS::BeginHostingNetworkDeprecated(Kernel::HLERequestContext& ctx) {
...
+    IPC::RequestBuilder rb = rp.MakeBuilder(1, 0);
+
+    // The host can not be kicked.
+    if (network_node_id == 1) {
+        rb.Push(ResultCode(ErrorDescription::NotAuthorized, ErrorModule::UDS,
+                           ErrorSummary::WrongArgument, ErrorLevel::Usage));
+        return;
+    }
+
+    std::lock_guard<std::mutex> lock(connection_status_mutex);
```
Comment (lioncash) [inline src/core/hle/service/nwm/nwm_uds.cpp:949]:
```suggestion
    std::lock_guard lock(connection_status_mutex);
```

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 3]
```diff
@@ -928,6 +931,57 @@ void NWM_UDS::BeginHostingNetworkDeprecated(Kernel::HLERequestContext& ctx) {
...
+    }
+
+    // This function always returns success if the status is valid.
+    rb.Push(RESULT_SUCCESS);
+
+    using Network::WifiPacket;
+    Network::MacAddress dest_address{};
+
+    if (network_node_id == BroadcastNetworkNodeId) {
+        // TODO(Subv): Send 3 deauth frames to the broadcast mac address
```
Comment (vvanelslande) [inline src/core/hle/service/nwm/nwm_uds.cpp:965]:
```suggestion
```

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 4]
```diff
@@ -928,6 +931,57 @@ void NWM_UDS::BeginHostingNetworkDeprecated(Kernel::HLERequestContext& ctx) {
...
+        }
+        dest_address = *address;
+    }
+
+    WifiPacket deauth;
+    deauth.channel = network_channel;
+    deauth.data = {};
+    deauth.destination_address = dest_address;
+    deauth.type = WifiPacket::PacketType::Deauthentication;
+    SendPacket(deauth);
```
Comment (vvanelslande) [inline src/core/hle/service/nwm/nwm_uds.cpp:982]:
```suggestion
    SendPacket(deauth);
    
    if (network_node_id == BroadcastNetworkNodeId) {
        SendPacket(deauth);
        SendPacket(deauth);
    }
```

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 5]
```diff
@@ -928,6 +931,57 @@ void NWM_UDS::BeginHostingNetworkDeprecated(Kernel::HLERequestContext& ctx) {
...
+        if (!address) {
+            // There is no error if the network node id was not found.
+            return;
+        }
+        dest_address = *address;
+    }
+
+    WifiPacket deauth;
+    deauth.channel = network_channel;
+    deauth.data = {};
```
Comment (vvanelslande) [inline src/core/hle/service/nwm/nwm_uds.cpp:979]: this line doesn't do anything
```suggestion
```

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 6]
```diff
@@ -928,6 +931,57 @@ void NWM_UDS::BeginHostingNetworkDeprecated(Kernel::HLERequestContext& ctx) {
...
+                           ErrorSummary::InvalidState, ErrorLevel::Usage));
+        LOG_WARNING(Service_NWM, "called with status {}", connection_status.status);
+        return;
+    }
+
+    // This function always returns success if the status is valid.
+    rb.Push(RESULT_SUCCESS);
+
+    using Network::WifiPacket;
+    Network::MacAddress dest_address{};
```
Comment (vvanelslande) [inline src/core/hle/service/nwm/nwm_uds.cpp:962]: `{}` doesn't do anything
```suggestion
    Network::MacAddress dest_address;
```

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 7]
```diff
@@ -279,6 +279,7 @@ void NWM_UDS::HandleEAPoLPacket(const Network::WifiPacket& packet) {
 
         // We're now connected, signal the application
         connection_status.status = static_cast<u32>(NetworkStatus::ConnectedAsClient);
+        connection_status.disconnect_reason = 1;
```
Comment (lioncash) [inline src/core/hle/service/nwm/nwm_uds.cpp:282]: Perhaps these magic values should be part of a named enum?

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 8]
```diff
@@ -837,6 +839,8 @@ ResultCode NWM_UDS::BeginHostingNetwork(const u8* network_info_buffer,
         ASSERT_MSG(network_info.max_nodes > 1, "Trying to host a network of only one member.");
 
         connection_status.status = static_cast<u32>(NetworkStatus::ConnectedAsHost);
+        connection_status.disconnect_reason = static_cast<u32>(DisconnectStatus::Connected);
+        ;
```
Comment (vvanelslande) [inline src/core/hle/service/nwm/nwm_uds.cpp:843]:
```suggestion
```

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 9]
```diff
@@ -928,6 +931,57 @@ void NWM_UDS::BeginHostingNetworkDeprecated(Kernel::HLERequestContext& ctx) {
...
+        }
+        dest_address = *address;
+    }
+
+    WifiPacket deauth;
+    deauth.channel = network_channel;
+    deauth.destination_address = dest_address;
+    deauth.type = WifiPacket::PacketType::Deauthentication;
+    SendPacket(deauth);
+    if (network_node_id == BroadcastNetworkNodeId) {
```
Comment (Subv) [inline src/core/hle/service/nwm/nwm_uds.cpp:979]: maybe add a comment of why this is sending more than one deauth frame

**Pull request merged**